File: kotlin/largest-series-product/src/main/kotlin/LargestSeriesProduct.kt
class Series(val input: String) {
    init {
        require(input.all { it.isDigit() })
    }
    fun getLargestProduct(span: Int) = when {
        span == 0 -> 1L
        span > input.length -> throw IllegalArgumentException()
        else -> input.map(Character::getNumericValue).windowed(span).maxOf {
            it.reduce { acc, n -> acc * n }.toLong()
        }
    }
}

File: fsharp/largest-series-product/LargestSeriesProduct.fs
module LargestSeriesProduct

open System

let largestProduct (input: string) len: int option =
    if not (Seq.forall Char.IsDigit input) || len < 0 || input.Length < len then
        None
    elif len = 0 then
        Some 1
    else
        input
        |> Seq.map (fun c -> c.ToString() |> int)
        |> Seq.windowed len
        |> Seq.map (Seq.reduce (*))
        |> Seq.max
        |> Some
